common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/EstsTelemetry.java

```diff
@@ -321,17 +363,63 @@ String getCurrentTelemetryHeaderString() {
     }
 
-    String getLastTelemetryHeaderString() {
-        RequestTelemetry lastTelemetry = loadLastRequestTelemetryFromCache();
-
-        if (lastTelemetry == null) {
+    private String getLastTelemetryHeaderString() {
+        if (mLastRequestTelemetryCache == null) {
             return null;
         }
 
-        return lastTelemetry.getCompleteTelemetryHeaderString();
+        LastRequestTelemetry lastRequestTelemetry = (LastRequestTelemetry) mLastRequestTelemetryCache.getRequestTelemetryFromCache();
+
+        if (lastRequestTelemetry == null) {
+            // we did not have anything in the telemetry cache for the last request
+            // let's create a new object based on the data available from the current request object
+            // and return the header string formed via that object
+            final String correlationId = DiagnosticContext.getRequestContext().get(DiagnosticContext.CORRELATION_ID);
+            final CurrentRequestTelemetry currentRequestTelemetry = mTelemetryMap.get(correlationId);
+            lastRequestTelemetry = new LastRequestTelemetry(currentRequestTelemetry.getSchemaVersion());
+            lastRequestTelemetry = (LastRequestTelemetry) lastRequestTelemetry.derive(currentRequestTelemetry);
+            return lastRequestTelemetry.getCompleteHeaderString();
+        }
+
+        // create a copy of the object retrieved from cache
+        LastRequestTelemetry lastRequestTelemetryCopy = new LastRequestTelemetry(lastRequestTelemetry.getSchemaVersion());
+        lastRequestTelemetryCopy = (LastRequestTelemetry) lastRequestTelemetryCopy.derive(lastRequestTelemetry);
+
+        // failed request data from the object retrieved from cache
+        final List<FailedRequest> originalFailedRequests = lastRequestTelemetry.getFailedRequests();
+
+        // error data from the object retrieved from cache
+        final List<String> originalErrors = lastRequestTelemetry.getErrors();
+
+        // the array index that marks how many elements were included in the header sent in the request
+        int indexOfFailedRequestsAndErrors = 0;
+
+        for (int i = 0; i < originalFailedRequests.size(); i++) {
+            // there is a limit of 8KB for the payload sent in request headers
+            // we will be maxing out at 4KB to avoid HTTP 413 errors
+            // check if we have enough space in the String to store another failed request/error element
+            // if yes, then add it to the failed request array (for the copy)
+            if (lastRequestTelemetryCopy.getCompleteHeaderString().length() < 3800) {
+                lastRequestTelemetryCopy.appendFailedRequestWithError(
+                        originalFailedRequests.get(indexOfFailedRequestsAndErrors),
+                        originalErrors.get(indexOfFailedRequestsAndErrors++)
+                );
+            } else {
+                // if there is no room for more data, then break out of this loop
+                break;
```

Comment:
Should we indicate the server somehow  that we are not sending the data due to size restrictions. i.e. just for us to be aware that this data could be skewed?

Reply:
Such a field is not currently part of the schema. I can add something to the platform telemetry for Android. Something like a 0 or 1 value indicating whether all the data is sent or not. So when we view our own telemetry then we know whether we have sent all the data or not. Or we could actually add a number indicating how many errors were either sent or kept on hold. I think the former is better i.e. it is probably sufficient to know whether all was sent or not and tell us to look at additional rows in the PerRequestTable. I don't see much value from knowing exactly how much was sent or not. What do you think @kreedula ?

Reply:
I think sending a boolean (0 or 1) to indicate if the data is partial or complete would be helpful .My primary concern raising this is , potentially in future we might end up sending partial data for every request and keep on accumulating old data , thus latest data could be lagging. Atleast having this information will help us to investigate and we probably can increase the payload if needed.

Reply:
Addressed